### PR TITLE
Modify the password and privilege checking logic

### PR DESCRIPTION
**Pull request by morningman**
**Password checking logic**

There are some problems with the current password and permission checking logic. For example:
First, we create a user by:
`create user cmy@"%" identified by "12345";`

And then 'cmy' can login with password '12345' from any hosts.

Second, we create another user by:
`create user cmy@"192.168.%" identified by "abcde";`

Because "192.168.%" has a higher priority in the permission table than "%". So when "cmy" try
to login in by password "12345" from host "192.168.1.1", it should match the second permission
entry, and will be rejected because of invalid password.
But in current implementation, Doris will continue to check password on first entry, than let it pass. So we should change it.

**Permission checking logic**

After a user login, it should has a unique identity which is got from permission table. For example,
when "cmy" from host "192.168.1.1" login, it's identity should be `cmy@"192.168.%"`. And Doris
should use this identity to check other permission, not by using the user's real identity, which is
`cmy@"192.168.1.1"`.

**Black list**
Functionally speaking, Doris only support adding WHITE LIST, which is to allow user to login from
those hosts in the white list. But is some cases, we do need a BLACK LIST function.
Fortunately, by changing the logic described above, we can simulate the effect of the BLACK LIST.

For example, First we add a user by:
`create user cmy@'%' identified by '12345';`

And now user 'cmy' can login from any hosts. and if we don't want 'cmy' to login from host A, we
can add a new user by:
`create user cmy@'A' identified by 'other_passwd';`

Because "A" has a higher priority in the permission table than "%". If 'cmy' try to login from A using password '12345', it will be rejected.

ISSUE #2205

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/http/rest/TableQueryPlanAction.java [hunk 1]
```diff
@@ -146,15 +145,8 @@ protected void executeWithoutPassword(ActionAuthorizationInfo authInfo, BaseRequ
                                                  "only support OlapTable currently, but Table [" + tableName + "] "
                                                     + "is not a OlapTable");
                 }
-                // construct ConnectContext  for analyzer
-                ConnectContext context = new ConnectContext(null);
-                context.setCatalog(Catalog.getCurrentCatalog());
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/http/rest/TableQueryPlanAction.java]: Is it OK that catalog and cluster is not set?

File: fe/src/main/java/org/apache/doris/mysql/privilege/UserProperty.java [hunk 2]
```diff
@@ -150,44 +148,18 @@ public boolean isSuperuser() {
...
-        for (Map.Entry<TablePattern, PrivBitSet> entry : privsMap.entrySet()) {
-            whiteList.revokePrivsFromDomain(domain, entry.getKey(), entry.getValue(),
-                                            errOnNonExist, true /* check */);
-        }
-
-        for (Map.Entry<TablePattern, PrivBitSet> entry : privsMap.entrySet()) {
-            whiteList.revokePrivsFromDomain(domain, entry.getKey(), entry.getValue(),
-                                            errOnNonExist, false /* check */);
-        }
+    public void removeDamain(String domain) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/mysql/privilege/UserProperty.java]:
```suggestion
    public void removeDomain(String domain) {
```

File: fe/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java [hunk 3]
```diff
@@ -52,28 +61,24 @@ public void addEntry(PrivEntry newEntry, boolean errOnExist, boolean errOnNonExi
...
-                        existingEntry.setSetByDomainResolver(false);
-                        LOG.info("merge priv entry: {}", existingEntry);
-                    }
-                    return;
+                checkOperationAllowed(existingEntry, newEntry, "ADD ENTRY");
+                if (existingEntry.isSetByDomainResolver() && newEntry.isSetByDomainResolver()) {
+                    existingEntry.setPrivSet(newEntry.getPrivSet());
+                    LOG.debug("reset priv entry: {}", existingEntry);
+                } else if (existingEntry.isSetByDomainResolver() && !newEntry.isSetByDomainResolver()
+                        || !existingEntry.isSetByDomainResolver() && !newEntry.isSetByDomainResolver()) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java]: seems that it equal with `!newEntry.isSetByDomainResolver()`

File: gensrc/thrift/Types.thrift [hunk 4]
```diff
@@ -372,3 +372,11 @@ enum TLoadSourceType {
     RAW,
     KAFKA,
 }
+
+// represent a user identity
+struct TUserIdentity {
+    1: required string username
+    2: required string host
+    3: required bool is_domain
```
[the remark this answers is not in the document]
Comment (morningman) [reply gensrc/thrift/Types.thrift]: OK

File: fe/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java (hunk 3 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/mysql/privilege/PrivTable.java]: You are right...

File: fe/src/main/java/org/apache/doris/http/rest/TableQueryPlanAction.java (hunk 1 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/http/rest/TableQueryPlanAction.java]: It OK, the ThreadLocal ConnectContext is already set before, with all necessary members.

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**